File: sdk/ash-client.js
/**
 * ash-client.js — ASH (Agent Standards Hub) JavaScript/Node 客户端 SDK
 *
 * 安装：直接 require 或 <script> 引入
 * 用法：
 *   const ASH = require('./ash-client.js');
 *   const client = new ASH('http://localhost:8933', 'ash_your_key');
 *   const card = await client.analyze('训练一个医疗问诊智能体');
 *
 * License: MIT
 */
'use strict';

const http = require('http');
const https = require('https');

class ASHClient {
  constructor(baseUrl, apiKey) {
    this.baseUrl = (baseUrl || 'http://localhost:8933').replace(/\/$/, '');
    this.apiKey = apiKey || '';
    this._fetch = baseUrl && baseUrl.startsWith('https') ? https : http;
  }

  _request(method, path, body) {
    return new Promise((resolve, reject) => {
      const url = new URL(this.baseUrl + path);
      const opts = {
        method,
        hostname: url.hostname,
        port: url.port || (url.protocol === 'https:' ? 443 : 80),
        path: url.pathname + url.search,
        headers: { 'Content-Type': 'application/json' },
        timeout: 30000,
      };
      if (this.apiKey) opts.headers['X-API-Key'] = this.apiKey;
      if (body) opts.headers['Content-Length'] = Buffer.byteLength(JSON.stringify(body));

      const req = this._fetch.request(opts, res => {
        let data = '';
        res.on('data', chunk => data += chunk);
        res.on('end', () => {
          try { resolve({ status: res.statusCode, data: JSON.parse(data) }); }
          catch (e) { resolve({ status: res.statusCode, data: { raw: data } }); }
        });
      });
      req.on('error', reject);
      req.on('timeout', () => { req.destroy(); reject(new Error('请求超时（30s）')); });
      if (body) req.write(JSON.stringify(body));
      req.end();
    });
  }


  /** 解包后端响应：{ok, data:{...}} → {...} ；{ok, ...} 原样返回 */
  _unwrap(r) {
    if (r.data && typeof r.data === 'object' && Object.keys(r.data).length === 2 && r.data.ok !== undefined && r.data.data !== undefined) {
      return r.data.data;
    }
    return r.data;
  }

    /* ===== 公共接口 ===== */

  /** 健康检查（无需 Key） */
  async health() {
    const r = await this._request('GET', '/api/health', null);
    return r.data;
  }

  /** 神经网络能力推荐（短名） */
  async recommend(industry, target) {
    const r = await this._request('POST', '/api/neural/recommend', {
      industry: industry || '', target: target || '对话助手',
    });
    return r.data;
  }

  /** 注册客户 → 返回 apiKey（兼容 register({name,contact}) 或 register(name,contact)） */
  async register(nameOrObj, contact) {
    let name, em;
    if (typeof nameOrObj === 'object' && nameOrObj) {
      name = String(nameOrObj.name || '匿名客户').slice(0, 60).trim();
      em = String(nameOrObj.contact || '').trim();
    } else {
      name = String(nameOrObj || '匿名客户').slice(0, 60).trim();
      em = String(contact || '').trim();
    }
    const r = await this._request('POST', '/api/auth/register', { name, contact: em });
    if (r.data.ok) this.apiKey = r.data.apiKey;
    return r.data;
  }

  /** 需求评估 → 需求卡 */
  async analyze(query, opts) {
    opts = opts || {};
    const r = await this._request('POST', '/api/engine/analyze', {
      query, industry: opts.industry || '', target: opts.target || '对话助手',
    });
    // 后端返回 {ok, id, card, ...}；同时兼容 {ok, data:{...}}
    return (r.data && r.data.card) ? r.data : (r.data.data || r.data);
  }

  /** 报价预估 */
  async quote(query, monthlyCalls, dataTokens) {
    const r = await this._request('POST', '/api/billing/quote', { query, monthlyCalls, dataTokens });
    return (r.data && r.data.monthlyCost !== undefined) ? r.data : (r.data.data || r.data);
  }

  /** 知识卡列表 */
  async listKnowledgeCards(opts) {
    const q = new URLSearchParams();
    if (opts && opts.category) q.set('category', opts.category);
    if (opts && opts.q) q.set('q', opts.q);
    const qs = q.toString();
    const r = await this._request('GET', '/api/knowledge-cards' + (qs ? '?' + qs : ''));
    return r.data;
  }

  async listCards() {
    const r = await this._request('GET', '/api/knowledge-cards');
    return (r.data && Array.isArray(r.data.cards)) ? r.data : (r.data.data || r.data);
  }

  /** 单张知识卡 */
  async getCard(packId) {
    const r = await this._request('GET', '/api/knowledge-card/' + packId);
    return r.data;
  }

  /** 能力包注册表 */
  async listPacks() {
    const r = await this._request('GET', '/api/packs');
    return r.data;
  }

  /** 神经能力推荐 */
  async neuralRecommend(industry) {
    const r = await this._request('POST', '/api/neural/recommend', { industry });
    return r.data;
  }

  /** 生成评估报告 */
  async generateReport(query, opts) {
    opts = opts || {};
    const r = await this._request('POST', '/api/report/pdf', {
      query, industry: opts.industry || '', target: opts.target || '对话助手', webhook: opts.webhook || '',
    });
    return r.data;
  }

  /** 计费标准 */
  async getPrices() {
    const r = await this._request('GET', '/api/billing/prices');
    return r.data;
  }

  /* ===== 客户接口（需 Key） ===== */

  /** 账户信息 */
  async me() {
    const r = await this._request('GET', '/api/auth/me');
    return r.data;
  }

  /** 用量明细 */
  async usage(limit) {
    const r = await this._request('GET', '/api/auth/usage?limit=' + (limit || 20));
    return r.data;
  }

  /** 订阅（年付 ¥3,900 含 2,000 万词元） */
  async subscribe(plan) {
    const r = await this._request('POST', '/api/auth/subscribe', { plan: plan || 'standard' });
    return r.data;
  }

  /** 充值（管理员手动入账） */
  async recharge(amount, method) {
    const r = await this._request('POST', '/api/auth/recharge', { amount: amount || 500, method: method || 'alipay' });
    return r.data;
  }

  /** 充值申请（客户提交，等运营确认） */
  async requestRecharge(amount, method) {
    const r = await this._request('POST', '/api/auth/recharge-request', { amount: amount || 500, method: method || 'alipay' });
    return r.data;
  }

  /** 撤销充值申请 */
  async cancelRecharge(orderId) {
    const r = await this._request('POST', '/api/auth/recharge-cancel', { orderId });
    return r.data;
  }

  /** 我的订单 */
  async orders() {
    const r = await this._request('GET', '/api/auth/orders');
    return r.data;
  }

  /** 我的报告 */
  async myReports() {
    const r = await this._request('GET', '/api/auth/reports');
    return r.data;
  }
}

/* 浏览器兼容 */
if (typeof window !== 'undefined') window.ASHClient = ASHClient;
/* Node.js 模块 */
if (typeof module !== "undefined") {
  const _exports = ASHClient;
  _exports.ASH = ASHClient;
  _exports.ASHClient = ASHClient;
  _exports.default = ASHClient;
  module.exports = _exports;
}

File: sdk/ash_client.py
#!/usr/bin/env python3
"""
ASH (Agent Standards Hub) Python 客户端 SDK

安装：直接 import 或 pip install ash-sdk（规划中）
用法：
    from ash_client import ASHClient
    client = ASHClient('http://localhost:8933', 'ash_your_key')
    card = client.analyze('训练一个医疗问诊智能体')

License: MIT
"""
import json
import urllib.request
import urllib.error
import urllib.parse
from typing import Optional, Dict, Any


class ASHClient:
    """ASH 智能体训练规范中心 — Python SDK"""

    def __init__(self, base_url: str = 'http://localhost:8933', api_key: str = ''):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key

    def _request(self, method: str, path: str, body: Optional[Dict] = None) -> Dict[str, Any]:
        url = self.base_url + path
        headers = {'Content-Type': 'application/json'}
        if self.api_key:
            headers['X-API-Key'] = self.api_key
        data = json.dumps(body).encode('utf-8') if body else None
        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                return json.loads(resp.read().decode('utf-8'))
        except urllib.error.HTTPError as e:
            try:
                return json.loads(e.read().decode('utf-8'))
            except Exception:
                return {'ok': False, 'error': f'HTTP {e.code}: {e.reason}'}
        except Exception as e:
            return {'ok': False, 'error': str(e)}

    # ===== 公共接口 =====

    def register(self, name_or_obj, contact: str = '') -> Dict:
        if isinstance(name_or_obj, dict):
            name = str(name_or_obj.get('name', '匿名客户'))[:60].strip()
            contact = (name_or_obj.get('contact', '') or contact or '').strip()
        else:
            name = str(name_or_obj or '匿名客户')[:60].strip()
        """注册客户 → 返回 apiKey（自动保存到实例）"""
        r = self._request('POST', '/api/auth/register', {'name': name, 'contact': contact})
        if r.get('ok') and r.get('apiKey'):
            self.api_key = r['apiKey']
        return r

    def analyze(self, query: str, industry: str = '', target: str = '对话助手') -> Dict:
        """需求评估（核心端点）→ 结构化需求卡"""
        return self._request('POST', '/api/engine/analyze',
                             {'query': query, 'industry': industry, 'target': target})

    def quote(self, query: str, monthly_calls: int = 10000, data_tokens: int = 0) -> Dict:
        """报价预估"""
        return self._request('POST', '/api/billing/quote',
                             {'query': query, 'monthlyCalls': monthly_calls, 'dataTokens': data_tokens})

    def list_cards(self) -> Dict:
        """知识卡列表（8 张）"""
        return self._request('GET', '/api/knowledge-cards')

    def list_knowledge_cards(self, category: str = '', q: str = '') -> Dict:
        """知识卡查询（分类/关键词筛选）"""
        qs = []
        if category: qs.append('category=' + urllib.parse.quote(category))
        if q: qs.append('q=' + urllib.parse.quote(q))
        path = '/api/knowledge-cards' + ('?' + '&'.join(qs) if qs else '')
        return self._request('GET', path)

    def health(self) -> Dict:
        """健康检查"""
        return self._request('GET', '/api/health')

    def get_card(self, pack_id: str) -> Dict:
        """单张知识卡（15 维度）"""
        return self._request('GET', f'/api/knowledge-card/{pack_id}')

    def list_packs(self) -> Dict:
        """能力包注册表"""
        return self._request('GET', '/api/packs')

    def neural_recommend(self, industry: str) -> Dict:
        """行业→神经能力推荐"""
        return self._request('POST', '/api/neural/recommend', {'industry': industry})

    def generate_report(self, query: str, industry: str = '', target: str = '对话助手',
                        webhook: str = '') -> Dict:
        """生成评估报告（MD 即时 + PDF 异步）"""
        payload = {'query': query, 'industry': industry, 'target': target}
        if webhook:
            payload['webhook'] = webhook
        return self._request('POST', '/api/report/pdf', payload)

    def get_prices(self) -> Dict:
        """计费标准"""
        return self._request('GET', '/api/billing/prices')

    # ===== 客户接口（需 Key） =====

    def me(self) -> Dict:
        """账户信息"""
        return self._request('GET', '/api/auth/me')

    def usage(self, limit: int = 20) -> Dict:
        """用量明细"""
        return self._request('GET', f'/api/auth/usage?limit={limit}')

    def subscribe(self, plan: str = 'standard') -> Dict:
        """订阅（年付 ¥3,900 含 2,000 万词元）"""
        return self._request('POST', '/api/auth/subscribe', {'plan': plan})

    def recharge(self, amount: float, method: str = 'alipay') -> Dict:
        """充值（管理员手动入账）"""
        return self._request('POST', '/api/auth/recharge', {'amount': amount, 'method': method})

    def request_recharge(self, amount: float, method: str = 'alipay') -> Dict:
        """充值申请（客户提交，等运营确认）"""
        return self._request('POST', '/api/auth/recharge-request', {'amount': amount, 'method': method})

    def cancel_recharge(self, order_id: str) -> Dict:
        """撤销充值申请"""
        return self._request('POST', '/api/auth/recharge-cancel', {'orderId': order_id})

    def orders(self) -> Dict:
        """我的订单"""
        return self._request('GET', '/api/auth/orders')

    def my_reports(self) -> Dict:
        """我的报告列表"""
        return self._request('GET', '/api/auth/reports')


if __name__ == '__main__':
    # 自测
    c = ASHClient('http://localhost:8933')
    print('=== 健康检查 ===')
    import urllib.request
    try:
        h = urllib.request.urlopen(c.base_url + '/api/health', timeout=5)
        print('OK' if h.status == 200 else f'HTTP {h.status}')
    except Exception as e:
        print(f'FAIL: {e}')

    print('\n=== 知识卡列表 ===')
    r = c.list_cards()
    if r.get('ok'):
        for card in r['cards']:
            print(f"  • {card.get('packId', '?')}: {card.get('definition', '')[:50]}...")
    else:
        print('  (服务不可用)')

# 别名：短名 ASH（同 JS 端 ASH 一致）
ASH = ASHClient
